src/ShellRouting/ShellRouter.cs

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Jupyter.Core.Protocol;
+
+namespace Microsoft.Jupyter.Core
+{
+
+    /// <summary>
+    ///     Routes shell messages to handlers based on the message type
+    ///     of each incoming shell messsge.
+    /// </summary>
+    public class ShellRouter : IShellRouter
+    {
+        private readonly IDictionary<string, Action<Message>> shellHandlers = new Dictionary<string, Action<Message>>();
+        private Action<Message> fallback;
+        private readonly ILogger<ShellRouter> logger;
+        private IServiceProvider services;
+
+        public ShellRouter(
+            IServiceProvider services,
+            ILogger<ShellRouter> logger
+        )
+        {
+            this.logger = logger;
+            this.services = services;
+
+            // Set a default fallback action.
+            RegisterFallback(message =>
+                logger.LogWarning("Unrecognized custom shell message of type {Type}: {Message}", message.Header.MessageType, message)
```

Comment:
can `logger` be null?
also, this is creating a capture on `logger`; it is probably a bit more efficient if you use `this.logger` instead.

Reply:
My understanding was that using `this.logger` would create a capture on `this` instead?

Reply:
With respect to nullability, it shouldn't be null as it comes from the dependency injection framework; I've added an explicit check to that effect as well as `#nullable enable` metadata to document that assumption progmatically. Thanks for pointing that out!